### PR TITLE
Implement indentation

### LINKED ISSUE #45: Indentation Not Working
Hello,

I've just installed and tried making use of `d2-mode` but indentation isn't working. If I type out the example
```
My Class {
   shape: class
   +field: string
   read ()
}
```
indentation behaves the same as in fundamental mode. I have to manually tab the `shape: class` line, and then I have to manually de-indent the curly brace. I've installed both from melpa and tried a git clone, but I get the same behavior. 

Are there limits on the indentation?

### PR DESCRIPTION
**Pull request by MoisMoshev**
Fixes #45 
Implements simple parsing of current and previous lines, and calculation of indentation, depending on the syntax.
It is still a bit rough, the regexes might have to be adjusted.

### CODE REVIEW

**Comment by andorsk** [comment]
@MoisMoshev thank you for the PR. I will review, and if things go well, merge it in. Thank you for the contribution!!!

**Comment by andorsk** [comment]
@MoisMoshev I am really swamped this week, but will look at it in more detail either later half of the week or this weekend. This is going to a major feature in the next version of the release. Thank you very much for contributing the PR!

**Comment by MoisMoshev** [comment]
Excellent, makes me very happy!
I'll be away next week (would probably be checking here in read-only mode) so I might not be able to react if intervention is needed on my side.

**Comment by andorsk** [comment]
Perfect. Thanks @MoisMoshev. Will get this merged in asap. Something came up this weekend and going to have to delay my review until mid next week. my apologies. I'm looking forward to the review.

File: d2-mode.el [hunk 1]
```diff
@@ -122,14 +122,161 @@
...
+                  (cons (cons (list (d2--decl-tag el))
+                              (cons (d2--decl-line el)
+                                    (d2--decl-column el)))
+                        acc)))
+              tokens
+              :initial-value ())))
+
+(defun d2--sort-tokens-by-line (tokens)
+  "Sort TOKENS by line number."
+  (sort tokens
```
Comment (andorsk) [inline d2-mode.el:162]: @MoisMoshev I ran this locally, and also ran into this problem. 3 arguments, but only accepts two. Can you clean this up?

File: d2-mode.el [hunk 2]
```diff
@@ -122,14 +122,161 @@
...
+                             (one-or-more space)))
+                   (? (group
+                       (one-or-more (not (any ?{ ?}))))))
+               'node))
+        (subnode-start (d2--locate-declaration
+                        (rx (group (one-or-more (any alnum "_" "<->"
+                                                     "->" "--" "<-"))
+                                   (? (group ":" (one-or-more space)))
+                                   (? (group (one-or-more (not (any ?{ ?})))))
+                                   "{"
```
Comment (andorsk) [inline d2-mode.el:187]:
```suggestion
                                   "{"))
```

File: d2-mode.el [hunk 3]
```diff
@@ -122,14 +122,161 @@
...
+                                   ))
```
Comment (andorsk) [inline d2-mode.el:188]:
```suggestion
```

File: d2-mode.el [hunk 4]
```diff
@@ -122,14 +122,161 @@
...
+                                     "<-")
+                                (? (seq
+                                    ":"
+                                    (one-or-more space)))
+                                (? (one-or-more graph))))
+                     'connection)))
+    (list node subnode-start end connection)))
+
+;;; declaration part handling
+(defun d2--decl-tag (decl) (car decl))
```
Comment (andorsk) [inline d2-mode.el:206]: TODO: Please update with doc strings

**Comment by andorsk** [comment]
@MoisMoshev firstly, thank you for the PR. I really appreciate it. I would love to get it in, but there are some problems. Would you mind taking a look at some of the comments and seeing if you could fix them?

- Andor

**Review by andorsk — CHANGES_REQUESTED** [review]
Thank you very much for the PR. Please fix comments below.

**Comment by andorsk** [comment]
@MoisMoshev once the build checks are passing, I'll be happy to re-evaluate this and try to merge it in. The PR was really helpful. Thank you again.

**Comment by MoisMoshev** [comment]
Thanks, I'll take care of it in the coming days.

File: d2-mode.el (hunk 1 above)
Comment (MoisMoshev) [reply d2-mode.el:162]: Hmm I'm confused by this, sort takes just one positional, and key arguments:
`(sort SEQ &key KEY LESSP REVERSE IN-PLACE)`
So it looks ok...? I don't think this changed recently. I'm running emacs 30.0.50

Comment (MoisMoshev) [reply d2-mode.el:162]: Nevermind, just checked in Emacs 29 and indeed the signature changed. It should be 
(sort SEQ PREDICATE)
brb

Comment (MoisMoshev) [reply d2-mode.el:162]: Fixed

**Comment by MoisMoshev** [comment]
Everything fixed

File: d2-mode.el (hunk 1 above)
Comment (andorsk) [reply d2-mode.el:162]: Love it. Thank you.

**Comment by andorsk** [comment]
> Everything fixed

Running everything over CI/CD. Then will check locally. If looks good, will merge in. Thank you!

**Comment by MoisMoshev** [comment]
Some seem like an issue in the environment, e.g.:
Error: Packages could not be installed:\n    Failed to verify signature: "let-alist-1.0.6.tar.sig"

This one I might have to fix:
[ELISP-CHECK] Running check: elisp-check-checkdoc
Error: Probably "matches" should be imperative "match"

**Review by andorsk — APPROVED** [review]
approved! ci/cd to be fixed before bumping tag on repo. Thanks for the PR! It was really helpful.

**Pull request merged**